File: src/leetcode/MinimumGeneticMutation.py
import sys
class Solution(object):
    def minMutation(self, start, end, bank):
        """
        :type start: str
        :type end: str
        :type bank: List[str]
        :rtype: int
        """
        if(start == end):
            return 0
        visited = []
        queue = []
        queue.append(start)
        queue.append("#")
        minValue = sys.maxint
        depth = 0
        while(queue):
            tmp = queue[0]
            queue.pop(0)
            if(tmp == "#"):
                depth += 1
                if(queue):
                    queue.append("#")
            elif(tmp == end):
                minValue = min(depth, minValue)
            else:
                for gene in bank:
                    if(solution.compareStrings(gene, tmp) and gene not in visited):
                        visited.append(gene)
                        queue.append(gene)
        return -1 if minValue==sys.maxint else minValue

    def compareStrings(self, gene, tmp):
        flag = 0
        for i in range(len(gene)):
            if not gene[i] == tmp[i]:
                flag += 1
        return flag if flag == 1 else 0


if __name__ == "__main__":
    solution = Solution()
    start = "AAAACCCC"
    end = "CCCCCCCC"
    bank = ["AAAACCCA","AAACCCCA","AACCCCCA","AACCCCCC","ACCCCCCC","CCCCCCCC","AAACCCCC","AACCCCCC"]
    result = solution.minMutation(start, end, bank)
    print result

File: src/leetcode/MinimumGeneticMutation.java
package leetcode;

import java.util.*;

public class MinimumGeneticMutation {
	public static int minMutation(String start, String end, String[] bank) {
		if(start.equals(end)) 
			return 0;		
		Set<String> visited = new HashSet<>();
		Queue<String> queue = new LinkedList<>();
		
		queue.offer(start);
		queue.offer("#");
		int depth = 0;
		int min = Integer.MAX_VALUE;
		while(!queue.isEmpty()) {
			String current = queue.poll();
			System.out.println(queue);
			System.out.println(min);
			if(current.equals("#")) {
				depth++;
				System.out.println(depth);
				if(!queue.isEmpty())
					queue.offer("#");
			}
			else if(current.equals(end)) 
			   min = Math.min(depth, min);
			else{
			    for(String gene: bank) {
				    if(compareStrings(current, gene) == 1 && !visited.contains(gene)) {
					    visited.add(gene);
					    queue.offer(gene);
				    }
			    }
			}
		}		
        return min==Integer.MAX_VALUE? -1 : min;
    }
	public static int compareStrings(String stringA, String stringB) {
		int len = 8;
		int flag = 0;
		for(int i=0; i<len; i++){
			if(stringA.charAt(i) != stringB.charAt(i))
				flag += 1;
		}
		if(flag == 1)
			return flag;
		else
			return -1;
	}
	public static void main(String[] args) {
		String start = "AAAACCCC";
		String end = "CCCCCCCC";
		String[] bank = new String[]{"AAAACCCA","AAACCCCA","AACCCCCA","AACCCCCC","ACCCCCCC","CCCCCCCC","AAACCCCC","AACCCCCC"};
		int result = minMutation(start, end, bank);
		System.out.println(result);
	}
}
